File: spec/open_current_spec.rb
require "date"
require "tempfile"

RSpec.describe "opening links with a Plug mapping" do

  it "from an obsidian link, one per line" do
    # given
    vim.command 'nmap zz <Plug>NotesOpenCurrent'
    vim.command 'set hidden'
    vim.insert '- This is [[The Target]]'
    vim.normal
    vim.feedkeys 'gg'

    # when
    vim.feedkeys 'zz'

    # then
    filename = vim.command 'echo @%'
    expect(filename).to eq('The Target.md')
  end

  it "from an obsidian link, two per line, not on link" do
    # given
    vim.command 'nmap zz <Plug>NotesOpenCurrent'
    vim.command 'set hidden'
    vim.insert '- a [[The Target]] a [[The Other Target]]'
    vim.normal
    vim.feedkeys 'gg'

    # when
    vim.feedkeys 'zz'

    # then
    filename = vim.command 'echo @%'
    expect(filename).to eq('')
  end

  it "from an obsidian link, two per line, first link" do
    # given
    vim.command 'nmap zz <Plug>NotesOpenCurrent'
    vim.command 'set hidden'
    vim.insert '- a [[The Target]] a [[The Other Target]]'
    vim.normal
    vim.feedkeys 'gglllllll'

    # when
    vim.feedkeys 'zz'

    # then
    filename = vim.command 'echo @%'
    expect(filename).to eq('The Target.md')
  end

  it "from an obsidian link, two per line, second link" do
    # given
    vim.command 'nmap zz <Plug>NotesOpenCurrent'
    vim.command 'set hidden'
    vim.insert '- a [[The Target]] a [[The Other Target]]'
    vim.normal
    vim.feedkeys 'ggllllllllllllllllllllllll'

    # when
    vim.feedkeys 'zz'

    # then
    filename = vim.command 'echo @%'
    expect(filename).to eq('The Other Target.md')
  end

  it "from a list item" do
    # given
    vim.command 'nmap zz <Plug>NotesOpenCurrent'
    vim.command 'set hidden'
    vim.insert '- The Target'
    vim.normal
    vim.feedkeys 'gg'

    # when
    vim.feedkeys 'zz'

    # then
    filename = vim.command 'echo @%'
    expect(filename).to eq('The Target.md')
  end

end

File: tests/open_current_spec.lua
local helpers = require('tests.support.helpers')

describe('opening links with NotesOpenCurrent', function()
  before_each(function()
    helpers.setup_test_env()
  end)

  after_each(function()
    helpers.teardown_test_env()
  end)

  -- Test using the new command directly
  it('command works from an obsidian link', function()
    -- given
    helpers.set_buffer_content('- This is [[The Target]]')
    vim.cmd('normal! gg')

    -- when
    vim.cmd('NotesOpenCurrent')

    -- then
    local filename = vim.fn.expand('%:t')
    assert.are.equal('The Target.md', filename)
  end)

  -- Test using the Lua function directly
  it('lua function works from an obsidian link', function()
    -- given
    helpers.set_buffer_content('- This is [[The Direct Target]]')
    vim.cmd('normal! gg')

    -- when
    require('notes').open_current()

    -- then
    local filename = vim.fn.expand('%:t')
    assert.are.equal('The Direct Target.md', filename)
  end)

  it('from an obsidian link, one per line', function()
    -- given
    helpers.set_buffer_content('- This is [[The Target]]')
    vim.cmd('normal! gg')

    -- when
    vim.cmd('NotesOpenCurrent')

    -- then
    local filename = vim.fn.expand('%:t')
    assert.are.equal('The Target.md', filename)
  end)

  it('from an obsidian link, two per line, not on link', function()
    -- given
    helpers.set_buffer_content('- a [[The Target]] a [[The Other Target]]')
    vim.cmd('normal! gg')

    -- when
    vim.cmd('NotesOpenCurrent')

    -- then
    local filename = vim.fn.expand('%:t')
    assert.are.equal('', filename)
  end)

  it('from an obsidian link, two per line, first link', function()
    -- given
    helpers.set_buffer_content('- a [[The Target]] a [[The Other Target]]')
    vim.cmd('normal! gglllllll')

    -- when
    vim.cmd('NotesOpenCurrent')

    -- then
    local filename = vim.fn.expand('%:t')
    assert.are.equal('The Target.md', filename)
  end)

  it('from an obsidian link, two per line, second link', function()
    -- given
    helpers.set_buffer_content('- a [[The Target]] a [[The Other Target]]')
    vim.cmd('normal! ggllllllllllllllllllllllll')

    -- when
    vim.cmd('NotesOpenCurrent')

    -- then
    local filename = vim.fn.expand('%:t')
    assert.are.equal('The Other Target.md', filename)
  end)

  it('does nothing when not on obsidian link', function()
    -- given
    helpers.set_buffer_content('- The Target')
    vim.cmd('normal! gg')

    -- when
    vim.cmd('NotesOpenCurrent')

    -- then
    local filename = vim.fn.expand('%:t')
    assert.are.equal('', filename)
  end)
end)
